### ISSUE TITLE
WiFi support for ESP32

### ISSUE THREAD

**Issue by zephyrbot** [zephyrproject-rtos/zephyr#3723]
**_Reported by Leandro Pereira:_**

WiFi support requires binary blobs that currently assumes it's running on FreeRTOS.  Figure out a way to use them on Zephyr.

(Imported from Jira ZEP-2288)

**Comment by locomuco** [comment]
for anybody who's interested, [here](https://github.com/locomuco/zephyr-esp32) is a application that can compile the ESP wifi lib against zephyr, but still quite a long way to go getting wifi running

**Comment by galak** [comment]
Adding reference to the #10539

**Comment by lpereira** [comment]
@galak #10539 is unrelated to @locomuco's application -- ESP8266 in that PR is being used as an external chip, running its stock firmware, rather than running Zephyr and driving the radio itself.

**Comment by mchodzikiewicz** [comment]
Hi, it's been a while... are there any updates on this?

**Comment by MarcDorner** [comment]
Hi, is there any update on the topic or further information how the status is?

**Comment by Deliver88** [comment]
Hello, does zephyr support the wifi on esp32? The sample/net/wifi is comparable for esp32 but after fleshing there is no interaction.

**Comment by JaimeAlbq** [comment]
Hello every one,

I tried to build the WiFi example, but I got the following error and note log:

```
FAILED: zephyr/drivers/serial/CMakeFiles/drivers__serial.dir/uart_esp32.c.obj 
ccache /home/user/.espressif/tools/xtensa-esp32-elf/esp-2019r2-8.2.0/xtensa-esp32-elf/bin/xtensa-esp32-elf-gcc -DBUILD_VERSION=v2.2.0-rc1-255-gb01cd4c7bbc3 -DKERNEL -D_FORTIFY_SOURCE=2 -D__ZEPHYR_SUPERVISOR__ -D__ZEPHYR__=1 -I/home/user/workspace/repositories/zephyr/include -Izephyr/include/generated -I/home/user/workspace/repositories/zephyr/soc/xtensa/esp32 -I/home/user/workspace/repositories/zephyr/soc/xtensa/esp32/include -I/home/user/workspace/repositories/zephyr/ext/lib/fnmatch/. -I/home/user/workspace/repositories/zephyr/modules/hal/esp-idf/zephyr/../components/esp32/include -I/home/user/workspace/repositories/zephyr/modules/hal/esp-idf/zephyr/../components/soc/esp32/include -I/home/user/workspace/repositories/zephyr/modules/hal/xtensa/include -I/home/user/workspace/repositories/zephyr/modules/hal/xtensa/zephyr/soc/esp32 -isystem /home/user/workspace/repositories/zephyr/lib/libc/minimal/include -isystem /home/user/.espressif/tools/xtensa-esp32-elf/esp-2019r2-8.2.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/8.2.0/include -isystem /home/user/.espressif/tools/xtensa-esp32-elf/esp-2019r2-8.2.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/8.2.0/include-fixed -Os -imacros/home/user/workspace/projects/examples/esp32/zephyr/wifi/build/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -imacros/home/user/workspace/repositories/zephyr/include/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-pointer-sign -Wpointer-arith -Wno-unused-but-set-variable -Werror=implicit-int -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-strict-overflow -fno-reorder-functions -fno-defer-pop -fmacro-prefix-map=/home/user/workspace/projects/examples/esp32/zephyr/wifi=CMAKE_SOURCE_DIR -fmacro-prefix-map=/home/user/workspace/repositories/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/home/user/workspace/repositories/zephyr=WEST_TOPDIR -ffunction-sections -fdata-sections -mlongcalls -std=c99 -nostdinc -MD -MT zephyr/drivers/serial/CMakeFiles/drivers__serial.dir/uart_esp32.c.obj -MF zephyr/drivers/serial/CMakeFiles/drivers__serial.dir/uart_esp32.c.obj.d -o zephyr/drivers/serial/CMakeFiles/drivers__serial.dir/uart_esp32.c.obj   -c /home/user/workspace/repositories/zephyr/drivers/serial/uart_esp32.c
/home/user/workspace/repositories/zephyr/drivers/serial/uart_esp32.c:458:2: error: request for member 'irq_config_func' in something not a structure or union
  .irq_config_func = uart_esp32_irq_config_func_##idx,
  ^
/home/user/workspace/repositories/zephyr/drivers/serial/uart_esp32.c:486:3: note: in expansion of macro 'ESP32_UART_IRQ_HANDLER_FUNC'
   ESP32_UART_IRQ_HANDLER_FUNC(idx)          \
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/user/workspace/repositories/zephyr/drivers/serial/uart_esp32.c:541:1: note: in expansion of macro 'ESP32_UART_INIT'
 ESP32_UART_INIT(0);
 ^~~~~~~~~~~~~~~
[113/166] Building C object zephyr/sub...es/subsys__net__ip.dir/net_shell.c.obj
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /usr/bin/cmake --build /home/user/workspace/projects/examples/esp32/zephyr/wifi/build
```

How to proceed?

Hopefully this help in some way.

**Comment by figual** [comment]
This is a bug (just a missing comma in line 485 of drivers/serial/uart_esp32.c, line 485 (end of line, missing a comma). I submitted a PR yesterday about it.

**Comment by JaimeAlbq** [comment]
Now I am facing the same problem as @Deliver88.

It can be built, thanks to @figual, but when I try to flash the `esptool.py` says it:
`A fatal error occurred: Invalid segment count 17 (max 16). Usually this indicates a linker script problem.`

So I tried to force to accept it by increasing the number of segments in the line 1603 in `esptool.py` from 16 to 17.

After that the board can be flash. But when I open the serial port I see it doesn't boot properly.

The log from serial:
```
ets Jun  8 2016 00:22:57

rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3ffb0000,len:180
load:0x3ffb00b4,len:92
load:0x3ffb0110,len:56
load:0x3ffb0148,len:96
load:0x3ffb01a8,len:20
ho 0 tail 12 room 4
load:0x3ffb01bc,len:40
ho 0 tail 12 room 4                                                            
load:0x3ffb01e4,len:64                                                         
load:0x3ffb0240,len:96                                                         
ho 0 tail 12 room 4                                                            
load:0x3ffb02a0,len:304                                                        
load:0x3ffb03d0,len:14380                                                      
load:0x40080000,len:1024                                                       
load:0x40080400,len:256                                                        
load:0x40080500,len:180                                                        
load:0x400805b4,len:184                                                        
load:0x4008066c,len:16                                                         
ho 0 tail 12 room 4                                                            
load:0x4008067c,len:72                                                         
ho 0 tail 12 room 4                                                            
load:0x400806c4,len:66008                                                      
1162 mmu set 00010000, pos 00010000                                            
entry 0x40081140
```

**Comment by JaimeAlbq** [comment]
I could workaround to flash the code by disabling the configuration option [`CONFIG_NET_LOG`](https://docs.zephyrproject.org/latest/reference/kconfig/CONFIG_NET_LOG.html#cmdoption-arg-config-net-log) from [`wifi/prj.conf`](https://github.com/zephyrproject-rtos/zephyr/blob/master/samples/net/wifi/prj.conf).

The shell works fine, and I really enjoy to used.

But whenever it tries to use some drive, such as Wi-Fi and I²C, the board get locked.

**Comment by jenmwms** [comment]
Adding reference to the #21492

**Comment by WilliamGFish** [comment]
> Adding reference to the #21492

HI Jen,
I think we are looking at two different things, the work being done under #21492 etc is a UART driven AT modem driver where as this is relating to using the esp as an MCU rather than a slave modem.

Billy..

**Comment by matteofacchinetti** [comment]
Hi,
I'm tryng to use wifi modem driver building  wifi example from the current mainline.
With menuconfig I needed to add WIFI_ESP support:
```
-TIMEOUT_64BIT y
 LEGACY_TIMEOUT_API n -> y
 MODEM n -> y
 NET_OFFLOAD n -> y
 WIFI_ESP n -> y
 WIFI_OFFLOAD n -> y
+MODEM_CMD_HANDLER y
+MODEM_CMD_HANDLER_MAX_PARAM_COUNT 6
+MODEM_CONTEXT y
+MODEM_CONTEXT_MAX_NUM 1
+MODEM_CONTEXT_VERBOSE_DEBUG n
+MODEM_GSM_PPP n
+MODEM_IFACE_UART y
+MODEM_RECEIVER n
+MODEM_SHELL n
+MODEM_SOCKET n
+MODEM_UBLOX_SARA n
+MODEM_WNCM14A2A n
+NET_DEFAULT_IF_OFFLOAD n
+NET_OFFLOAD_LOG_LEVEL 0
+WIFI_ESP_AT_VERSION_1_7 n
+WIFI_ESP_AT_VERSION_2_0 y
+WIFI_ESP_NAME "esp-wifi-modem"
+WIFI_ESP_PASSIVE_TCP n
+WIFI_ESP_RX_STACK_SIZE 1024
+WIFI_ESP_RX_THREAD_PRIORITY 7
+WIFI_ESP_WORKQ_STACK_SIZE 2048
+WIFI_ESP_WORKQ_THREAD_PRIORITY 7
```
 
but I got the following error in  zephyr/drivers/wifi/esp/esp.c:

```
In file included from ../include/arch/xtensa/arch.h:18:0,
                 from ../include/arch/cpu.h:27,
                 from ../include/kernel_includes.h:39,
                 from ../include/kernel.h:17,
                 from /workdir/zephyrproject/zephyr/drivers/wifi/esp/esp.c:13:
/workdir/zephyrproject/zephyr/drivers/wifi/esp/esp.c: In function 'esp_init':
../include/devicetree.h:225:40: error: 'DT_N_INST_0_espressif_esp_BUS_P_label' undeclared (first use in this function)
 #define DT_INST(inst, compat) UTIL_CAT(DT_N_INST, DT_DASH(inst, compat))
                                        ^
../include/devicetree.h:1526:38: note: in definition of macro 'DT_CAT'
 #define DT_CAT(node_id, prop_suffix) node_id##prop_suffix
                                      ^
../include/devicetree.h:1080:31: note: in expansion of macro 'DT_PROP'
 #define DT_BUS_LABEL(node_id) DT_PROP(DT_BUS(node_id), label)
                               ^
../include/devicetree.h:1073:25: note: in expansion of macro 'DT_CAT'
 #define DT_BUS(node_id) DT_CAT(node_id, _BUS)
                         ^
../include/devicetree.h:1080:39: note: in expansion of macro 'DT_BUS'
 #define DT_BUS_LABEL(node_id) DT_PROP(DT_BUS(node_id), label)
                                       ^
../include/devicetree.h:1391:33: note: in expansion of macro 'DT_BUS_LABEL'
 #define DT_INST_BUS_LABEL(inst) DT_BUS_LABEL(DT_DRV_INST(inst))
                                 ^
../include/sys/util.h:421:26: note: in expansion of macro 'UTIL_PRIMITIVE_CAT'
 #define UTIL_CAT(a, ...) UTIL_PRIMITIVE_CAT(a, __VA_ARGS__)
                          ^
../include/devicetree.h:225:31: note: in expansion of macro 'UTIL_CAT'
 #define DT_INST(inst, compat) UTIL_CAT(DT_N_INST, DT_DASH(inst, compat))
                               ^
../include/devicetree.h:1148:27: note: in expansion of macro 'DT_INST'
 #define DT_DRV_INST(inst) DT_INST(inst, DT_DRV_COMPAT)
                           ^
../include/devicetree.h:1391:46: note: in expansion of macro 'DT_DRV_INST'
 #define DT_INST_BUS_LABEL(inst) DT_BUS_LABEL(DT_DRV_INST(inst))
                                              ^
/workdir/zephyrproject/zephyr/drivers/wifi/esp/esp.c:843:9: note: in expansion of macro 'DT_INST_BUS_LABEL'
         DT_INST_BUS_LABEL(0));
         ^
../include/devicetree.h:225:40: note: each undeclared identifier is reported only once for each function it appears in
 #define DT_INST(inst, compat) UTIL_CAT(DT_N_INST, DT_DASH(inst, compat))
                                        ^
../include/devicetree.h:1526:38: note: in definition of macro 'DT_CAT'
 #define DT_CAT(node_id, prop_suffix) node_id##prop_suffix
                                      ^
../include/devicetree.h:1080:31: note: in expansion of macro 'DT_PROP'
 #define DT_BUS_LABEL(node_id) DT_PROP(DT_BUS(node_id), label)
                               ^
../include/devicetree.h:1073:25: note: in expansion of macro 'DT_CAT'
 #define DT_BUS(node_id) DT_CAT(node_id, _BUS)
                         ^
../include/devicetree.h:1080:39: note: in expansion of macro 'DT_BUS'
 #define DT_BUS_LABEL(node_id) DT_PROP(DT_BUS(node_id), label)
                                       ^
../include/devicetree.h:1391:33: note: in expansion of macro 'DT_BUS_LABEL'
 #define DT_INST_BUS_LABEL(inst) DT_BUS_LABEL(DT_DRV_INST(inst))
                                 ^
../include/sys/util.h:421:26: note: in expansion of macro 'UTIL_PRIMITIVE_CAT'
 #define UTIL_CAT(a, ...) UTIL_PRIMITIVE_CAT(a, __VA_ARGS__)
                          ^
../include/devicetree.h:225:31: note: in expansion of macro 'UTIL_CAT'
 #define DT_INST(inst, compat) UTIL_CAT(DT_N_INST, DT_DASH(inst, compat))
                               ^
../include/devicetree.h:1148:27: note: in expansion of macro 'DT_INST'
 #define DT_DRV_INST(inst) DT_INST(inst, DT_DRV_COMPAT)
                           ^
../include/devicetree.h:1391:46: note: in expansion of macro 'DT_DRV_INST'
 #define DT_INST_BUS_LABEL(inst) DT_BUS_LABEL(DT_DRV_INST(inst))
                                              ^
/workdir/zephyrproject/zephyr/drivers/wifi/esp/esp.c:843:9: note: in expansion of macro 'DT_INST_BUS_LABEL'
         DT_INST_BUS_LABEL(0));
         ^
[92/145] Building C object zephyr/subsys/net/ip/CMakeFiles/subsys__net__ip.dir/net_if.c.obj
ninja: build stopped: subcommand failed.
```

I don't understand exactly what's happen. Is there anything missing in DTS? 
Any help is welcome.

**Comment by Skallwar** [comment]
Hi. I'm interested on working on this, but I have no idea where to start from. Have you some resources or previous work to help start working on it ?

**Comment by chrta** [comment]
@Skallwar Since the ESP32 wifi driver is implemented in RIOT, the code there might be a good starting point: https://github.com/RIOT-OS/RIOT/pull/9426/files

**Comment by Skallwar** [comment]
I've worked a bit on this and I have a dummy Wifi driver. Right now, ESP-IDF Wifi code is not really reusable, but on the [v4.1-beta](https://github.com/espressif/esp-idf/releases/tag/v4.1-beta2) there is a [file](https://github.com/espressif/esp-idf/blob/c5cb44d8135a5ec6678259e30468632bc2442288/components/esp_wifi/esp32/esp_adapter.c) with set of stub functions that could be adapted to Zephyr.
I don't know the ETA for this change on a stable release however. When this change will be stable how do I proceed ? I create a pull request to the [zephyr esp-idf repo](https://github.com/zephyrproject-rtos/esp-idf) to update it ?

**Comment by lpereira** [comment]
The main reason WiFi isn't supported in Zephyr on the ESP32 is not just the
driver.  Zephyr runs entirely on the IRAM, loaded by the first stage
bootloader in ROM.  If the driver+libs were to be linked, it wouldn't fit.
You'd need to enable the Flash Cache, which will require a second stage
bootloader (because the code will be overwritten when it's done). There was
some code to do this back when I was working on it, and some people
actually made it work, but I don't know what's the status of it. There
might be some code in some branch in my fork.

**Comment by Skallwar** [comment]
> There was some code to do this back when I was working on it, and some people actually made it work, but I don't know what's the status of it

Are we talking of [this PR](https://github.com/zephyrproject-rtos/zephyr/pull/8783) ?

**Comment by lpereira** [comment]
That seems to be it, yeah.

Once this is working -- you probably want to add ESP32 support in mcuboot,
too, to act as the second stage bootloader -- it should be trivial to hook
up the ESP32 WiFi blob with Zephyr's network stack, IIRC.

With some work it should be possible to deconstruct the static library and
write shims to use the crypto API (and select which crypto implementation
to use).

**Comment by Skallwar** [comment]
> That seems to be it, yeah.

Any idea why this has never been merged ?

> Once this is working -- you probably want to add ESP32 support in mcuboot, too, to act as the second stage bootloader

The esp-idf bootloader is not sufficient ?

> it should be trivial to hook up the ESP32 WiFi blob with Zephyr's network stack, IIRC. 

Is it really a blob ? From what I see [here](https://github.com/espressif/esp-idf/tree/v4.1-beta2/components/esp_wifi) it looks like no, but you maybe referring to something else ?

**Comment by lpereira** [comment]
The ESP-IDF bootloader could be used as a starting point, yes, but I think that the ideal situation for Zephyr would be using mcuboot so that the Zephyr tooling could be used.

You have to link with the blobs provided in https://github.com/espressif/esp32-wifi-lib (the "lib" directory is a git submodule pointing to this repo).

**Comment by locomuco** [comment]
@Skallwar 

https://github.com/locomuco/zephyr-esp32

this was the follow up to the previously mentioned PR, it was able to enable flash cache and link in the Wifi Blob, it's outdated, but might give good hints in the single commits

**Comment by Skallwar** [comment]
> The ESP-IDF bootloader could be used as a starting point, yes

I can not boot the hello world example with it, the boot is stuck at ```boot: Disabling RNG early entropy source...``` after loading the app. Beside on larger project, the app contains too much segments (here segment = elf section) and the app will not be loaded.

I will maybe try to work on mcuboot

**Comment by mahavirj** [comment]
Update:

Feature branch that has functional networking (WiFi) on ESP32 can be found [here](https://github.com/mahavirj/zephyr/tree/feature/esp32_networking_support_on_zephyr_v2.4.0-rc1).

Technical details:

 - This is still based on 2nd stage bootloader from ESP-IDF which does required flash cache setup for XIP enabled application
 - This feature branch has appropriate `west` manifest file to pick up correct base ESP-IDF repository (based on `v4.2` version)
 - Toolchain should be based on revision `xtensa-esp32-elf-gcc (crosstool-NG esp-2020r2) 8.2.0`
 - All networking application should be functional, just that `prj.conf` should be added with credentials for WiFi network using `CONFIG_WIFI_SSID="myssid"` and `CONFIG_WIFI_PASSWORD="mypassword"`
 
Official PR (from Espressif) will still take some time but interim please feel free to give this a try and report any issues or feedback.

CC @ExtremeGTX @Skallwar @GautierAtWork

**Comment by lpereira** [comment]
Great job, Mahavir!

**Comment by chrta** [comment]
@mahavirj I tried to test your branch but failed, because the path to the source files in https://github.com/mahavirj/zephyr/blob/feature/esp32_networking_support_on_zephyr_v2.4.0-rc1/drivers/wifi/esp32/CMakeLists.txt#L69 did not match the correct paths in my esp-idf module (I used west and one manual `git submodule update --init` in esp-idf module). Maybe i did something wrong.
Could you describe the steps necessary to test your branch (especially how to get the correct esp-idf and what variables to set)?

**Comment by mahavirj** [comment]
@chrta

>  Could you describe the steps necessary to test your branch (especially how to get the correct esp-idf and what variables to set)?

Please find instructions [here](https://github.com/mahavirj/zephyr/wiki/ESP32-Getting-Started-Guide). Let me know if you face any issues.

**Comment by chrta** [comment]
> Please find instructions here. Let me know if you face any issues.

Great, thank you. Now i could proceed a little bit further. My previous error was that i used the wrong hal (the old one not named espressif).
Now i am facing the next issue with the python modules that are required (when i execute "west build"). I am using Ubuntu 20.04 and the build step fails because the necessary python modules are not installed. I try to install them from the requirements.txt file from the hal, but that did not work either.
Can you add instructions how to install the required python modules, or what venv to use when executing the "west build" step?

Maybe you can enable the issues in your github fork and we can continue the discussion there? I do not want to spam this issue too much.

**Comment by mahavirj** [comment]
> Maybe you can enable the issues in your github fork and we can continue the discussion there?

Done. I have also reached out you on slack channel, in case that is faster.

**Comment by chrta** [comment]
Just for reference: It works now, after fixing some python issues on my pc and applying https://github.com/espressif/esp-idf/commit/96cfdf3184f5fac4c744fa9b8bc76a3e69fe2b65 to esp-idf. Thank you @mahavirj for your support.

**Comment by timdesi** [comment]
@mahavirj i could confirm that WiFi on ESP32 works with your branch. Thx.

**Comment by vtothsvk** [comment]
Hi @mahavirj
I was trying out the WiFi support for ESP32 and encountered the following issue: each time I build an app and try to flash it I get an error pointing that no bootloader.bin is present in the build (also no partitions file is generated as well). Opened an [issue](https://github.com/mahavirj/zephyr/issues/1) on your repo as well about it.

**Comment by ExtremeGTX** [comment]
@vtothsvk I have this issue also on Windows 10, but you can use Linux for now to try the wifi support.
I will try to find a fix for this issue.

**Comment by mahavirj** [comment]
@ExtremeGTX FYI, fix for this is available [here](https://github.com/mahavirj/zephyr/issues/1#issuecomment-730866500). We will ensure that this is part of PR for WiFi support on ESP32.

**Comment by beriberikix** [comment]
@mahavirj thanks for the updates! Is there an official PR yet for us to follow?

**Comment by mahavirj** [comment]
@beriberikix Sorry for delayed reply, baseline PR is here https://github.com/zephyrproject-rtos/zephyr/pull/30424. Subsequent work will be based on this, it is waiting on more approvals and final merge.

**Comment by Nithyasrikannathal** [comment]
> > Please find instructions here. Let me know if you face any issues.
> 
> Great, thank you. Now i could proceed a little bit further. My previous error was that i used the wrong hal (the old one not named espressif). Now i am facing the next issue with the python modules that are required (when i execute "west build"). I am using Ubuntu 20.04 and the build step fails because the necessary python modules are not installed. I try to install them from the requirements.txt file from the hal, but that did not work either. Can you add instructions how to install the required python modules, or what venv to use when executing the "west build" step?
> 
> Maybe you can enable the issues in your github fork and we can continue the discussion there? I do not want to spam this issue too much.




how did you resolve the python isuue?

**Comment by Nithyasrikannathal** [comment]
> @chrta
> 
> > Could you describe the steps necessary to test your branch (especially how to get the correct esp-idf and what variables to set)?
> 
> Please find instructions [here](https://github.com/mahavirj/zephyr/wiki/ESP32-Getting-Started-Guide). Let me know if you face any issues.

I am facing a new issue which says that "  C:/Users/NFI2KOR/zephyr3wifi/modules/hal/espressif/components/esp_timer/src/esp_timer_impl_frc_legacy.c:141:1: error: unknown type name 'portMUX_TYPE'
 portMUX_TYPE s_time_update_lock = portMUX_INITIALIZER_UNLOCKED;
 ^~~~~~~~~~~~
C:/Users/NFI2KOR/zephyr3wifi/modules/hal/espressif/components/esp_timer/src/esp_timer_impl_frc_legacy.c:141:35: error: 'portMUX_INITIALIZER_UNLOCKED' undeclared here (not in a function)
 portMUX_TYPE s_time_update_lock = portMUX_INITIALIZER_UNLOCKED;
                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~" 
I gave west update . It gave error
" updating hal_espressif (modules/hal/espressif):
--- hal_espressif: fetching, need revision master"
fatal: couldn't find remote ref master
HEAD is now at 143df6755 CMakeLists.txt: make BL2 configurable
ERROR: update failed for project hal_espressif

Initially when i got this error i just cloned the [https://github.com/mahavirj/hal_espressif](url) inside the espressif folder. but i guess there is a issue

**Comment by WilliamGFish** [comment]
@Nithyasrikannathal 
You may find that if you refresh the repo the issue goes away, there has been a number of changes in the Wi-Fi area.
Billy..

**Comment by Nithyasrikannathal** [comment]
> @Nithyasrikannathal You may find that if you refresh the repo the issue goes away, there has been a number of changes in the Wi-Fi area. Billy..

Thanks for the reply. Can you please link the changes made or the updated thing? ?

**Comment by WilliamGFish** [comment]
> > @Nithyasrikannathal You may find that if you refresh the repo the issue goes away, there has been a number of changes in the Wi-Fi area. Billy..
> 
> Thanks for the reply. Can you please link the changes made or the updated thing? ?

If you look at the recently closed pull requests re wifi; you will see a number of changes to the shell which has had a knock on effect with many wifi drivers.
